modules/gateways/veritrans.php

```diff
@@ -349,21 +349,65 @@ function veritrans_link($params)
+        $whmcsApiCommand = 'GetInvoice';
+        $whmcsApiPostParams = array(
+            'invoiceid' => $invoiceId
+        );
+        $invoice = localAPI($whmcsApiCommand, $whmcsApiPostParams);
+        $existingInvoiceNotes = $invoice->notes;
     } catch (Exception $e) {
-        // error_log('Caught exception: ',  $e->getMessage(), "\n");
-        if (preg_match('/utilized/',$e->getMessage()) ){
-            return "<h4> Awaiting Your Payment </h4>";
+        $existingInvoiceNotes = "";
+    }
+
+    $midtransNotesSeparatorBegin = "#PayUrl: ";
+    $midtransNotesSeparatorEnd = " PayUrl#";
+
+    // Check if Snap URL/token found in invoices notes
+    if (strpos($existingInvoiceNotes, $midtransNotesSeparatorBegin)) {
+        // use Regexp to extract existing Snap Url value between Midtrans invoice notes begin & end separator
+        $midtransNotesRegexPattern = "/$midtransNotesSeparatorBegin(.*?)$midtransNotesSeparatorEnd/";
+        if ( preg_match($midtransNotesRegexPattern, $existingInvoiceNotes, $matches) ) {
+            $redirect_url = $matches[1];
+        }
+        // get Snap Token from string after the last `/` char of $redirect_url
+        $SnapUrlExplodeResults = explode('/', $redirect_url);
+        $snapToken = end( $SnapUrlExplodeResults );
+    }
+
+    // Check if Snap token not found from invoice notes
+    if( !isset($snapToken) ){
+        // Get snap token
+        try {
+            // debug
+            // echo "<pre>";
+            // var_dump($params);
+            // echo "</pre>";
+            // exit();
+            $snap_transaction = Veritrans_Snap::createTransaction($params);
+            $snapToken = $snap_transaction->token;
+            $redirect_url = $snap_transaction->redirect_url;
+            // error_log(" ############# TOKEN ::: ".$snapToken);
+
+            // save snap redirect url to invoice notes
+            try {
+                $whmcsApiCommand = 'UpdateInvoice';
+                $whmcsApiPostParams = array(
+                    'invoiceid' => $invoiceId,
+                    'notes' => $existingInvoiceNotes . $midtransNotesSeparatorBegin . $redirect_url . $midtransNotesSeparatorEnd                );
+                localAPI($whmcsApiCommand, $whmcsApiPostParams);
+            } catch (Exception $e) {
+                echo "error: unable to update invoice notes";
+            }
+
+        } catch (Exception $e) {
+            // error_log('Caught exception: ',  $e->getMessage(), "\n");
+            if (preg_match('/utilized|digunakan/', $e->getMessage())) {
+                // Display payment instruction, prevent payment button from being shown
+                return "Invoice ID has been created on Midtrans previously. Please try to check your email for the previous payment instruction details or open link on Invoice Notes.";
+            } else {
+                return "Oops, something went wrong when requested Payment token. Error:" . $e->getMessage();
+            }
```

Comment:
Small grammar improvement, should be `requesting` not `requested`. Or a better messaging should be:
`Midtrans payment module encountered unexpected error when requesting to Snap API. Error message:`